[AArch64] Add better fcmp costs for expanded predicates

davemgreen opened the pull request:
Certain fcmp predicates need to be expanded into multiple operations and or'd together. This adds some more accurate cost modelling for them based on the predicate. Unsupported operations are given the cost of a libcall and the latency is set to 2 as that seemed to be fairly common between different CPUs.

See https://llvm.godbolt.org/z/PsEsbfozG / https://llvm.godbolt.org/z/19anrn7xb / https://llvm.godbolt.org/z/PPYv3xooE for example codegen, and some SVE compares were improve with #145473.

davemgreen commented:
Rebase and ping - thanks.

SamTebbs33 commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4308,63 +4307,80 @@ InstructionCost AArch64TTIImpl::getCmpSelInstrCost(
+    if (LT.second.getScalarType() != MVT::f64 &&
+        LT.second.getScalarType() != MVT::f32 &&
+        LT.second.getScalarType() != MVT::f16)
+      return LT.first * getCallInstrCost(/*Function*/ nullptr, ValTy,
+                                         {ValTy, ValTy}, CostKind);
```
Do we have to account for the complexity of the libcall itself as well, or is the call instruction cost enough?

nasherm reviewed (commented):
This seems reasonable to me but I'd like to have another pair of eyes on it before approving

nasherm reviewed (approved):
LGTM

davemgreen committed:
[AArch64] Add better fcmp costs for expanded predicates

Certain fcmp predicates need to be expanded into multiple operations and or'd
together. This adds some more accurate core modelling for them based on the
predicate. Unsupported operations are given the cost of a libcall and the
latency is set to 2 as that seemed to be fairly common between different CPUs.
